### PR TITLE
Integration ACL check with saved objects.

### PR DESCRIPTION
**Pull request by SuZhou-Joe**
### Description

1. Make saved objects search under ACL DSL control.
2. Implement permission_service.

### Check List

- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
  - [ ] `yarn test:ftr`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [x] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: src/core/server/saved_objects/permission_control/client.ts [hunk 1]
```diff
@@ -38,54 +57,70 @@ export class SavedObjectsPermissionControl {
...
-        result: true,
+        result: hasAllPermission,
       };
     }
 
     return {
-      success: true,
-      result: false,
+      success: false,
+      error: 'Can not find target saved objects.',
```
Comment (wanglam) [inline src/core/server/saved_objects/permission_control/client.ts:88]: Do we need to use `i18n` wrap it?

File: src/core/server/saved_objects/permission_control/client.ts [hunk 2]
```diff
@@ -38,54 +57,70 @@ export class SavedObjectsPermissionControl {
...
+      return {
+        ...total,
+        [current.id]: new ACL(current.permissions).transformPermissions(),
+      };
+    }, {});
   }
 
   public async getPermittedWorkspaceIds(
     request: OpenSearchDashboardsRequest,
     permissionModeOrModes: SavedObjectsPermissionModes
```
Comment (wanglam) [inline src/core/server/saved_objects/permission_control/client.ts:107]: Could we rename this variable to `permissionModes`?

File: src/core/server/saved_objects/permission_control/client.ts [hunk 3]
```diff
@@ -38,54 +57,70 @@ export class SavedObjectsPermissionControl {
...
+    const queryDSL = ACL.genereateGetPermittedSavedObjectsQueryDSL(
+      permissionModeOrModes,
+      principals,
+      [WORKSPACE_TYPE]
+    );
+    const repository = this.createScopedRepository?.(request);
+    try {
+      const result = await repository?.find({
+        type: [WORKSPACE_TYPE],
+        queryDSL,
```
Comment (wanglam) [inline src/core/server/saved_objects/permission_control/client.ts:119]: Do we need to pass a size parameter to get all workspace ids?

File: src/core/server/saved_objects/permission_control/client.ts [hunk 4]
```diff
@@ -38,54 +57,70 @@ export class SavedObjectsPermissionControl {
...
+      };
+    }, {});
   }
 
   public async getPermittedWorkspaceIds(
     request: OpenSearchDashboardsRequest,
     permissionModeOrModes: SavedObjectsPermissionModes
   ) {
-    return [];
+    const principals = this.getPrincipalsFromRequest(request);
```
Comment (wanglam) [inline src/core/server/saved_objects/permission_control/client.ts:109]: Will we get all workspaces if principals is `{}`?

File: src/core/server/saved_objects/permission_control/client.ts [hunk 5]
```diff
@@ -14,21 +18,36 @@ export type SavedObjectsPermissionControlContract = Pick<
...
+    return payload;
   }
   private async bulkGetSavedObjects(
     request: OpenSearchDashboardsRequest,
     savedObjects: SavedObjectsBulkGetObject[]
   ) {
-    return (
-      (await this.getScopedSavedObjectsClient(request)?.bulkGet(savedObjects))?.saved_objects || []
-    );
+    return (await this.getScopedRepository(request)?.bulkGet(savedObjects))?.saved_objects || [];
```
Comment (wanglam) [inline src/core/server/saved_objects/permission_control/client.ts:47]: Why do we need to use scoped repository here? As I remember we need to change to the super user for general saved objects call.

File: src/core/server/workspaces/saved_objects/workspace_saved_objects_client_wrapper.ts [hunk 6]
```diff
@@ -187,26 +198,45 @@ export class WorkspaceSavedObjectsClientWrapper {
...
+          [PermissionMode.LibraryRead, PermissionMode.LibraryWrite, PermissionMode.Management]
+        );
+        if (options.workspaces) {
+          const isEveryWorkspaceIsPermitted = options.workspaces.every((item) =>
+            // TODO modify this line to use permittedWorkspaceIds if public workspace is also a workspace
+            ['public', ...(permittedWorkspaceIds || [])]?.includes(item)
+          );
+          if (!isEveryWorkspaceIsPermitted) {
+            throw generateWorkspacePermissionError();
+          }
```
Comment (wanglam) [inline src/core/server/workspaces/saved_objects/workspace_saved_objects_client_wrapper.ts:222]: Do we need to add `queryDSL` options here?

Comment (SuZhou-Joe) [reply src/core/server/workspaces/saved_objects/workspace_saved_objects_client_wrapper.ts:222]: I do not think we should. Either users have permission on workspace or permission on objects can get access to the objects, and in this if clause, users should access objects under workspace context.

File: src/core/server/saved_objects/permission_control/client.ts (hunk 5 above)
Comment (SuZhou-Joe) [reply src/core/server/saved_objects/permission_control/client.ts:47]: Agree, change that.

File: src/core/server/saved_objects/permission_control/client.ts (hunk 3 above)
Comment (SuZhou-Joe) [reply src/core/server/saved_objects/permission_control/client.ts:119]: Good catch, I will add a perPage to `999`

File: src/core/server/saved_objects/permission_control/client.ts (hunk 2 above)
Comment (SuZhou-Joe) [reply src/core/server/saved_objects/permission_control/client.ts:107]: Sure, done.

File: src/core/server/saved_objects/permission_control/client.ts (hunk 1 above)
Comment (SuZhou-Joe) [reply src/core/server/saved_objects/permission_control/client.ts:88]: We should, done.

File: src/core/server/saved_objects/permission_control/client.ts (hunk 4 above)
Comment (SuZhou-Joe) [reply src/core/server/saved_objects/permission_control/client.ts:113]: We will, if principals is `{}`, the queryDSL generated by ACL won't contain any ACL filter logic.

Comment (wanglam) [reply src/core/server/saved_objects/permission_control/client.ts:113]: Is that a correct logic? Seems it means the user have permission to all workspaces.

Comment (SuZhou-Joe) [reply src/core/server/saved_objects/permission_control/client.ts:113]: I get your point, let me adjust some code here.

**Review by wanglam — APPROVED** [review]

**Review by raintygao — APPROVED** [review]

**Pull request merged**